x/staking/keeper/msg_server.go

```diff
@@ -381,3 +384,112 @@ func (k msgServer) Undelegate(goCtx context.Context, msg *types.MsgUndelegate) (
 		CompletionTime: completionTime,
 	}, nil
 }
+
+// CancelUnbondingDelegation defines a method for canceling the unbonding delegation
+// and delegate back to validator
+func (k msgServer) CancelUnbondingDelegation(goCtx context.Context, msg *types.MsgCancelUnbondingDelegation) (*types.MsgCancelUnbondingDelegationResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	valAddr, err := sdk.ValAddressFromBech32(msg.ValidatorAddress)
+	if err != nil {
+		return nil, err
+	}
+	delegatorAddress, err := sdk.AccAddressFromBech32(msg.DelegatorAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	bondDenom := k.BondDenom(ctx)
+	if msg.Amount.Denom != bondDenom {
+		return nil, sdkerrors.Wrapf(
+			sdkerrors.ErrInvalidRequest, "invalid coin denomination: got %s, expected %s", msg.Amount.Denom, bondDenom,
+		)
+	}
+
+	// get validator
+	validator, found := k.GetValidator(ctx, valAddr)
+	if !found {
+		return nil, types.ErrNoValidatorFound
+	}
+
+	// In some situations, the exchange rate becomes invalid, e.g. if
+	// Validator loses all tokens due to slashing. In this case,
+	// make all future delegations invalid.
+	if validator.InvalidExRate() {
+		return nil, types.ErrDelegatorShareExRateInvalid
+	}
+
+	if validator.IsJailed() {
+		return nil, types.ErrValidatorJailed
+	}
+
+	ubd, found := k.GetUnbondingDelegation(ctx, delegatorAddress, valAddr)
+	if !found {
+		return nil, status.Errorf(
+			codes.NotFound,
+			"unbonding delegation with delegator %s not found for validator %s",
+			msg.DelegatorAddress, msg.ValidatorAddress)
+	}
+
+	var (
+		unbondEntry      types.UnbondingDelegationEntry
+		unbondEntryIndex int64 = -1
+	)
+
+	for i, entry := range ubd.Entries {
+		if entry.CreationHeight == int64(msg.CreationHeight) {
+			unbondEntry = entry
+			unbondEntryIndex = int64(i)
+			break
+		}
+	}
+	if unbondEntryIndex == -1 {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrNotFound, "unbonding delegation entry is not found at block height %d", msg.CreationHeight)
+	}
+
+	if !msg.Amount.IsPositive() {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "invalid amount")
+	}
+
+	if unbondEntry.Balance.LT(msg.Amount.Amount) {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "amount is greater than the unbonding delegation entry balance")
+	}
+
+	if ubd.Entries[unbondEntryIndex].CompletionTime.Before(ctx.BlockTime()) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "unbonding delegation is already processed.")
+	}
+
+	// delegate the unbonding delegation amount to validator back
+	_, err = k.Keeper.Delegate(ctx, delegatorAddress, msg.Amount.Amount, types.Unbonding, validator, false)
```

Comment:
Should we not call `Delegate` _after_ we modify the entry?

Reply:
We can do both ways @alexanderbez 
First update or delete the entry and call `Delegate` 
or 
Call `Delegate` first and update or delete the entry in unbonding delegations